python/pyspark/ml/torch/distributor.py

```diff
@@ -0,0 +1,491 @@
+
+    def __init__(
+        self,
+        framework: str,
+        num_processes: int = 1,
+        local_mode: bool = True,
+        use_gpu: bool = True,
+        spark: Optional[SparkSession] = None,
+    ):
+        """Initializes the distributor.
+
+        Parameters
+        ----------
+        framework : str
+            A string indicating whether or not we are using PyTorch or PyTorch
+            Lightning. This could either be the string “pytorch” or ”pytorch-lightning”.
+        num_processes : int, optional
+            An integer that determines how many different concurrent
+            tasks are allowed. We expect spark.task.gpus = 1 for GPU-enabled training. Default
+            should be 1; we don't want to invoke multiple cores/gpus without explicit mention.
+        local_mode : bool, optional
+            A boolean that determines whether we are using the driver
+            node for training. Default should be false; we don't want to invoke executors without
+            explicit mention.
+        use_gpu : bool, optional
+            A boolean that indicates whether or not we are doing training
+            on the GPU. Note that there are differences in how GPU-enabled code looks like and
+            how CPU-specific code looks like.
+        spark : Optional[SparkSession], optional
+            An optional parameter that allows users to pass in a custom SparkSession argument
+            with a custom conf, by default None
+
+        Raises
+        ------
+        ValueError
+            If any of the parameters are incorrect.
+        """
+        super().__init__(num_processes, local_mode, use_gpu, spark)
+        self.framework = framework
+        self.ssl_conf = "pytorch.spark.distributor.ignoreSsl"  # type: ignore
+        self._validate_input_params()
+
+    def _validate_input_params(self) -> None:
+        """Validates input params
+
+        Raises
+        ------
+        ValueError
+            Thrown when user fails to provide correct input params
+        """
+        super()._validate_input_params()
+        if self.framework not in self.available_frameworks:
+            raise ValueError(
+                f"{self.framework} is not a valid framework."
+                f"Available frameworks: {self.available_frameworks}"
+            )
+
+    def _run_local_training(
+        self,
+        framework_wrapper_fn: Optional[Callable],
+        train_fn: Union[Callable, str],
+        *args: Any,
+    ) -> Optional[Any]:
+        CUDA_VISIBLE_DEVICES = "CUDA_VISIBLE_DEVICES"
+        old_cuda_visible_devices = os.environ.get(CUDA_VISIBLE_DEVICES, "")
+        cuda_state_was_set = CUDA_VISIBLE_DEVICES in os.environ
+        if not framework_wrapper_fn:
+            raise RuntimeError("Unknown combination of parameters")
+        try:
+            if self.use_gpu:
+                if "gpu" not in self.sc.resources:
+                    raise ValueError('"gpu" was not found in the context resources.')
```

Comment:
If GPU is not available, could we just run on CPU instead of throwing an error?

Reply:
But we would need to inform the user that setting `use_gpu=True` lead to that issue. Alternatively, we could tell the user that gpu was not found so we will switch to cpu through logging,

Reply:
I prefer raise error in this case, and tell user to set use_gpu=False because no GPU configured.

Reply:
We actually did this on this [line](https://github.com/apache/spark/blob/28a6efffc84555d20db4aa036d847cfad7d3e98a/python/pyspark/ml/torch/distributor.py#L164) (from the previous PR) so we don't need this check anymore.